restapi/src/main/java/io/stargate/web/resources/Db.java

```diff
@@ -124,9 +95,35 @@ private DocumentDB getDocDataStoreForTokenInternal(TokenAndHeaders tokenAndHeade
         getAuthorizationService());
   }
 
-  public AuthenticationSubject getAuthenticationSubjectForToken(TokenAndHeaders tokenAndHeaders)
+  private AuthenticatedDB getRestDataStoreForTokenInternal(TokenAndHeaders tokenAndHeaders)
+      throws UnauthorizedException {
+
+    AuthenticationSubject authenticationSubject =
+        authenticationService.validateToken(tokenAndHeaders.token, tokenAndHeaders.headers);
+    DataStore dataStore =
+        dataStoreFactory.create(
+            authenticationSubject.asUser(),
+            DataStoreOptions.builder()
+                .alwaysPrepareQueries(true)
+                .putAllCustomProperties(tokenAndHeaders.headers)
+                .build());
+    return new AuthenticatedDB(dataStore, authenticationSubject);
+  }
+
+  public AuthenticatedDB getRestDataStoreForToken(String token, Map<String, String> headers)
       throws UnauthorizedException {
-    return authenticationService.validateToken(tokenAndHeaders.token, tokenAndHeaders.headers);
+    if (token == null) {
+      throw new UnauthorizedException("Missing token");
+    }
+
+    try {
+      return restTokensToDataStore.get(TokenAndHeaders.create(token, headers));
```

Comment:
I wonder if using _all_ headers a part of the cache key can prevent hits... if one of the headers is irrelevant, but changes its value often, will it effectively make the cache ineffective?

Reply:
I see that `TokenAndHeaders` filter out all but the `Host` header, but conceptually only the `AuthenticationService` knows that the relevant headers are... Should we somehow delegate the filtering of headers to `AuthenticationService`  for the sake of correctness?

Reply:
By similar logic, should the `AuthenticationService` control the cache duration? Otherwise, there's risk of continuing to allow a user to execute requests after the token expires. WDYT?

Reply:
Probably a good idea. I'll create a new issue.

Reply:
https://github.com/stargate/stargate/issues/915